### PR TITLE
Introduces function to determine allocated bytes for any value

### PR DESCRIPTION
**Pull request by Garanas**
This is useful to understand the impact of tables (or values in general) - a long missing factor in our hunt towards better performance. Looking forward to feedback on this function.

### CODE REVIEW

File: lua/sim/Unit.lua [hunk 1]
```diff
@@ -96,6 +96,45 @@ SyncMeta = {
...
+---@return integer
+function ToBytes(element, ignore, seen)
+
+    -- has no allocated bytes
+    if element == nil then
+        return 0
+    end
+
+    -- applies to tables and strings, to prevent counting them multiple times
+    local seen = seen or { }
```
Comment (4z0t) [inline lua/sim/Unit.lua:112]: It is local already

File: lua/sim/Unit.lua [hunk 2]
```diff
@@ -216,6 +255,14 @@ Unit = Class(moho.unit_methods) {
     OnCreate = function(self)
         local bp = self:GetBlueprint()
 
+        ForkThread(
+            function()
+                WaitSeconds(1.0)
+                LOG(ToBytes(self, { Blueprint = true, Brain = true, Audio = true, MovementEffects = true }))
+                LOG(ToBytes(__blueprints))
+            end
+        )
```
Comment (4z0t) [inline lua/sim/Unit.lua:264]: Test code.

File: lua/sim/Unit.lua [hunk 3]
```diff
@@ -96,6 +96,45 @@ SyncMeta = {
...
+    -- at this point we know it is a table
+    seen[element] = true
+    for k, v in element do
+        if not ignore[k] then
+            allocatedSize = allocatedSize + ToBytes(v, ignore, seen)
+        end
+    end
+
+    return allocatedSize
+end
```
Comment (4z0t) [inline lua/sim/Unit.lua:135]: Duplicate

**Review by 4z0t — CHANGES_REQUESTED** [review]

File: lua/sim/Unit.lua (hunk 3 above)
Comment (Garanas) [reply lua/sim/Unit.lua:135]: Test code too 😄

File: lua/sim/Unit.lua (hunk 1 above)
Comment (Garanas) [reply lua/sim/Unit.lua:112]: Parameters are, I'll adjust it

**Comment by Garanas** [comment]
Fun fact with the navigational mesh, on the map `Syrtis Major` the entire datastructure takes up about 33 megabytes:

image

On Seton's Clutch this is about 48 megabytes. This means that the compression we do works quite well.

About 30% of the memory that we use originates from these lines:
- https://github.com/FAForever/fa/blob/feature/bytes-function/lua/sim/NavGenerator.lua#L652-L670

If we remove those lookup tables:
- Syrtis Major: 33mb -> 21mb
- Seton's Clutch: 48mb -> 30mb

Oof 😄

**Comment by KionX** [comment]
This should exec only for tables: https://github.com/FAForever/fa/pull/4523/files#diff-b7cb9a852b1ea5bdbe28863255f535c0924f4446962075087db5bbe9a9b71314R54

**Comment by Garanas** [comment]
As far as I can see it only executes that for tables a this moment

**Comment by KionX** [comment]
if type(v) == 'table' then

**Comment by Garanas** [comment]
I'm all up for an alternative if it is more accurate!

As far as I can tell, the recursion only applies to tables at the moment, unless you're making a distinction between the hash and array sections of a table.

**Pull request merged**